docs: add bundling dependencies post (#46)
e18e/e18e: docs/blog/bundling-dependencies.md

---
title: Bundling dependencies (and when not to do it)
author:
  - name: The e18e Contributors
sidebar: false
date: 2025-02-10
head:
  - - meta
    - property: og:type
      content: website
  - - meta
    - property: og:title
      content: Bundling dependencies (and when not to do it)
  - - meta
    - property: og:url
      content: https://e18e.dev/blog/bundling-dependencies
  - - meta
    - property: og:description
      content: A brief write up on when you should or shouldn't bundle dependencies
---

# Bundling dependencies (and when not to do it)

As part of [e18e](https://e18e.dev), the community is often asked when you should or shouldn't bundle your dependencies _as a library author_.

> Much of this has come around thanks to people noticing that [vite]() does indeed bundle some of its dependencies. Similarly, [storybook]() has been known to do the same but for internal packages.

Reviewer: ```suggestion
Much of this has come around thanks to people noticing that [vite]() does indeed bundle most of its dependencies. Similarly, [storybook]() has been known to do the same but for internal packages.
```

- `@org/utils` - a library _only_ `@org/library` uses
- `@org/cli` - a cli package published to npm
- `@org/text` - a library used by `@org/library` and `@org/cli`

As you can see, `@org/utils` could be bundled into `@org/library` rather than being its own package. Meanwhile, `@org/text` is used by multiple published packages so should likely be published itself too.

**Advice:** If only used in one place, this is fine to bundle. Otherwise, publish the packages.

### Being "dependency free" / "zero dependency"

Especially with e18e community raising awareness of dependency bloat (amongst many other things), people have been adding more focus to going "dependency free".

In some projects, this makes a lot of sense for various reasons. For example, using more of what the platform provides rather than dependencies.

However, some projects have seen this (especially the ones losing users because they have bloated dependencies), and tried to "solve" the problem (trick people) by bundling _all_ of their dependencies.

These projects often include a `vendor/` directory of their npm packages, or run a bundler at pre-publish time.

This is really poor and such projects should be avoided.

> **Advice:** trying to trick people isn't the best approach. Put the work in and become dependency-free properly (if it really should be).

Author: you're right. I was being harsh 😅 

i'll try re-word while still pointing out it is very bad

This isn't great, but neither is the alternative (bundling the parts you use).

If possible, you really should try find a more appropriate package or create one which provides the part of the functionality you need.

### Extracting new packages

Sometimes, a package has a piece of functionality you need but that accounts for a tiny amount of the overall size.

If this is the case, it often makes sense to extract it into its own package.

However, the line is _very fine_ here.

If you extract something too little, you're publishing micro-utilities (which we're trying to avoid).

Ideally, you should extract _groups_ of functionality into re-usable packages. A good example of this is [empathic](https://github.com/lukeed/empathic), a group of path utilities.

### De-duplication

Whatever you do, if you bundle, you're at risk of removing the benefits npm gives of de-duplication.

Ideally, we want to declare our dependencies so we can leave npm to de-dupe and share them across other libraries being used in parallel to ours.

## Conclusion

> There's usually no good reason to bundle your dependencies.

Author: I touched on this a bit in that you had the option (but not the time presumably) to clean up those dependencies yourself (as a team).

I guess it was easier at the time to bundle them instead of doing the work to sort them out

that will still be the case for a lot of maintainers though, so I'll try reword

> There's usually no good reason to bundle your dependencies.

Author: can you take another look? 👀

> There's usually no good reason to bundle your dependencies.

Reviewer: looks good, thanks!

- `@org/cli` - a cli package published to npm
- `@org/text` - a library used by `@org/library` and `@org/cli`

As you can see, `@org/utils` could be bundled into `@org/library` rather than being its own package. Meanwhile, `@org/text` is used by multiple published packages so should likely be published itself too.

**Advice:** If only used in one place, this is fine to bundle. Otherwise, publish the packages.

### Being "dependency free" / "zero dependency"

Especially with e18e community raising awareness of dependency bloat (amongst many other things), people have been adding more focus to going "dependency free".

In some projects, this makes a lot of sense for various reasons. For example, using more of what the platform provides rather than dependencies.

However, some projects have seen this rise in demand for "dependency free", and tried to shortcut a solution by bundling _all_ of their dependencies.

This just hides the fact that they still have a bloated dependency tree, and worsens things by removing npm's ablity to de-dupe dependencies.

These projects often include a `vendor/` directory of their npm packages, or run a bundler at pre-publish time.

> This is really poor since it will hide the problem, but not solve it. So it is best to avoid these projects if there are alternatives available.

Author: this is a difficult point to make

some packages don't necessarily believe in cleaning up dependency trees but have noticed there is demand for it, so their "quick fix" has been to bundle all their dependencies in the hope it'll satisfy the demand (without having to do the actual work of writing less bloated code).

in those cases, a package still has all of the old, overly complex code internally but doesn't pull it from npm. this doesn't solve any problem.

on the other hand, there are projects which may believe in this effort and have chosen to bundle dependencies they can't change from (yet). but they generally wouldn't bundle _everything_, only the ones they're stuck with for now and aim to replace later

> This is really poor since it will hide the problem, but not solve it. So it is best to avoid these projects if there are alternatives available.

Author: I've also reworded this slightly, if you could re-check it

### Extracting new packages

Sometimes, a package has a piece of functionality you need but that accounts for a tiny amount of the overall size.

If this is the case, it often makes sense to extract it into its own package.

However, the line is _very fine_ here.

If you extract something too little, you're publishing micro-utilities (which we're trying to avoid).

Ideally, you should extract _groups_ of functionality into re-usable packages. A good example of this is [empathic](https://github.com/lukeed/empathic), a group of path utilities.

### De-duplication

Whatever you do, if you bundle, you're at risk of removing the benefits npm gives of de-duplication.

Ideally, we want to declare our dependencies so we can leave npm to de-dupe and share them across other libraries being used in parallel to ours.

## Real-world examples

### vite

vite currently bundles most of its dependencies as a way to tree-shake things up front so its consumers don't have to.

Given that it is a developer tool, unused in production, this made sense at the time it was decided.

> However, dependencies have vastly improved since then and the community has been hard at work improving performance across the board. These days, it is likely the vite team would not choose to bundle most of their dependencies.

Reviewer: This sounds a bit too optimistic. Last time we discussed this option with the team, there was heavy push back. I don't know when are we going to be unbundling vite deps, and if we do it will probably happen gradually.
```suggestion
However, dependencies have vastly improved since then and the community has been hard at work improving performance across the board. It is likely that the vite team will stop bundling some of their dependencies in the future.
```
